dotnet/src/Connectors/Connectors.Memory.Pinecone/PineconeUtils.cs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.SemanticKernel.Connectors.Memory.Pinecone.Http.ApiSchema;
+using Microsoft.SemanticKernel.Connectors.Memory.Pinecone.Model;
+
+namespace Microsoft.SemanticKernel.Connectors.Memory.Pinecone;
+
+public static class PineconeUtils
+{
+    /// <summary>
+    ///  The maximum size of the metadata associated with each vector.
+    ///  See https://docs.pinecone.io/docs/metadata-filtering#supported-metadata-size
+    /// </summary>
+    private const int MaxMetadataSize = 40 * 1024;
+    
+    /// <summary>
+    ///  The default dimension for Pinecone vectors. Equivalent to text-embeddings-ada-002 dimension.
+    /// </summary>
+    public const int DefaultDimension = 1536;
+    
+    public const string DefaultIndexName = "sk-index";
+    
+    /// <summary>
+    /// Defaults to cosine similarity.
+    /// </summary>
+    public const IndexMetric DefaultIndexMetric = IndexMetric.Cosine;
+    
+    /// <summary>
+    ///  The standard index type for Pinecone vectors.
+    /// </summary>
+    public const PodType DefaultPodType = PodType.P1X1;
+
+    internal static JsonSerializerOptions DefaultSerializerOptions => new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        WriteIndented = true,
+        AllowTrailingCommas = false,
+        ReadCommentHandling = JsonCommentHandling.Skip,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+        UnknownTypeHandling = JsonUnknownTypeHandling.JsonNode,
+        NumberHandling = JsonNumberHandling.AllowReadingFromString,
+        Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
+        Converters =
+        {
+            new PodTypeJsonConverter(),
+            new JsonStringEnumConverter(JsonNamingPolicy.CamelCase)
+        }
+    };
+
+    /// <summary>
+    ///  Utility method to ensure that the metadata size is not too large.
+    ///  This is necessary because Pinecone has a limit on the size of the metadata
+    ///  associated with each vector.
+    /// </summary>
+    /// <param name="documents"></param>
+    /// <returns> A stream of documents with valid metadata. </returns>
+    /// <remarks>
+    ///  If the metadata size is too large, the document will be split into multiple documents.
+    ///  The text field will be removed from the metadata and the document ID will be set to {document ID}-{split counter}
+    ///  The text field will be added to the metadata of the split documents. The split documents will be returned in the
+    ///  same order as the original document.
+    /// </remarks>
+    internal static async IAsyncEnumerable<PineconeDocument> EnsureValidMetadataAsync(
```

Comment:
Let's add unit tests for this method. We should also verify the performance of this method in case of big text size.